src/Altinn.ApiClients.Maskinporten/Service/MaskinportenService.cs

```diff
@@ -0,0 +1,227 @@
+{
+    public class MaskinportenService: IMaskinporten
+    {
+        private readonly HttpClient _client;
+
+        private readonly MaskinportenSettings _maskinportenConfig;
+
+        private readonly ILogger _logger;
+
+        private readonly IClientSecret _clientSecret;
+
+        private readonly IMemoryCache _memoryCache;
+
+        public MaskinportenService(HttpClient httpClient, 
+            IOptions<MaskinportenSettings> maskinportenConfig, 
+            ILogger<MaskinportenService> logger,
+            IMemoryCache memoryCache,
+            IClientSecret clientSecret)
+        {
+            httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            _client = httpClient;
+            _maskinportenConfig = maskinportenConfig.Value;
+            _logger = logger;
+            _memoryCache = memoryCache;
+            _clientSecret = clientSecret;
+        }
+
+     
+        public async Task<TokenResponse> GetToken(X509Certificate2 cert, string clientId, string scope, string resource, bool disableCaching = false)
+        {
+            return await GetToken(cert, null, clientId, scope, resource, disableCaching);
+        }
+
+        public async Task<TokenResponse> GetToken(JsonWebKey jwk, string clientId, string scope, string resource, bool disableCaching = false)
+        {
+            return await GetToken(null, jwk, clientId, scope, resource, disableCaching);
+        }
+
+        public async Task<TokenResponse> GetToken(string base64EncodedJwk, string clientId, string scope, string resource, bool disableCaching = false)
+        {
+            byte[] base64EncodedBytes = Convert.FromBase64String(base64EncodedJwk);
+            string jwkjson = Encoding.UTF8.GetString(base64EncodedBytes);
+            JsonWebKey jwk = new JsonWebKey(jwkjson);
+            return await GetToken(null, jwk, clientId, scope, resource, disableCaching);
+        }
+
+        public async Task<TokenResponse> GetToken(bool disableCaching = false)
+        {
+            ClientSecrets clientSecrets = await _clientSecret.GetClientSecrets();
+            if (clientSecrets.ClientKey != null)
+            {
+                return await GetToken(null, clientSecrets.ClientKey, _maskinportenConfig.ClientId, _maskinportenConfig.Scope, _maskinportenConfig.Resource, disableCaching);
+            }
+            else
+            {
+                return await GetToken(clientSecrets.ClientCertificate, null , _maskinportenConfig.ClientId, _maskinportenConfig.Scope, _maskinportenConfig.Resource, disableCaching);
+            }
+        }
+
+        private async Task<TokenResponse> GetToken(X509Certificate2 cert, JsonWebKey jwk, string clientId, string scope, string resource, bool disableCaching)
+        {
+            string cacheKey = $"{clientId}-{scope}";
+            TokenResponse accesstokenResponse;
+            if (disableCaching || !_memoryCache.TryGetValue(cacheKey, out accesstokenResponse))
```

Comment:
Do we need some sort of locking here to avoid making multiple requests to Maskinporten when the token is missing or expires?

Reply:
Added semaphore. Add to put it outside cachetry.